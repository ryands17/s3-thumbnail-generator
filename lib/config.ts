import { config } from 'dotenv'
config()

export const REGION = process.env.REGION || 'us-east-2'
