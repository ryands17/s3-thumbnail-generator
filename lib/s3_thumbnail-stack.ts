import * as cdk from '@aws-cdk/core'
import * as logs from '@aws-cdk/aws-logs'
import * as eventSource from '@aws-cdk/aws-lambda-event-sources'
import * as sqs from '@aws-cdk/aws-sqs'
import * as lambda from '@aws-cdk/aws-lambda'
import * as s3 from '@aws-cdk/aws-s3'
import * as s3Notif from '@aws-cdk/aws-s3-notifications'

const bucketName = 'all-images-bucket'

export class S3ThumbnailStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    // Queue to handle the S3 event source
    const queue = new sqs.Queue(this, 'thumbnailQueue', {
      queueName: 'thumbnailPayload',
      receiveMessageWaitTime: cdk.Duration.seconds(20),
      visibilityTimeout: cdk.Duration.seconds(60),
      retentionPeriod: cdk.Duration.days(2),
    })

    // Lambda that will react to S3 events
    const handler = new lambda.Function(this, 'resizeImage', {
      runtime: lambda.Runtime.NODEJS_12_X,
      code: lambda.Code.fromAsset('resources'),
      handler: 'index.handler',
      timeout: cdk.Duration.seconds(10),
      memorySize: 256,
      reservedConcurrentExecutions: 20,
      logRetention: logs.RetentionDays.ONE_WEEK,
      environment: {
        QUEUE_URL: queue.queueUrl,
        REGION: props?.env?.region!,
      },
    })

    // Attach the queue as an event source to the Lambda
    handler.addEventSource(new eventSource.SqsEventSource(queue))

    // S3 bucket for storing images
    const imagesBucket = new s3.Bucket(this, 'allImagesBucket', {
      bucketName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    })

    // Attach the queue to react to S3 object creates
    imagesBucket.addObjectCreatedNotification(
      new s3Notif.SqsDestination(queue),
      {
        prefix: 'photos/',
      }
    )
  }
}
