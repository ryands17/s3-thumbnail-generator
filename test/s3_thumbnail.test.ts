import * as cdk from '@aws-cdk/core'
import {
  SynthUtils,
  expect as expectCDK,
  haveResourceLike,
} from '@aws-cdk/assert'
import { S3ThumbnailStack } from '../lib/s3_thumbnail-stack'

const stackName = 'S3ThumbnailStack'

// Commented due to an issue with snapshots changing in every run

// test('snapshot works correctly', () => {
//   const app = new cdk.App()
//   const stack = new S3ThumbnailStack(app, stackName)
//   expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
// })

test('S3, SQS and Lambda resources are created', () => {
  const app = new cdk.App()
  const stack = new S3ThumbnailStack(app, stackName)

  expectCDK(stack).to(
    haveResourceLike('AWS::S3::Bucket', {
      BucketName: 'all-images-bucket',
    })
  )

  expectCDK(stack).to(
    haveResourceLike('AWS::SQS::Queue', {
      MessageRetentionPeriod: 172800,
      QueueName: 'thumbnailPayload',
      ReceiveMessageWaitTimeSeconds: 20,
      VisibilityTimeout: 60,
    })
  )

  expectCDK(stack).to(
    haveResourceLike('AWS::Lambda::Function', {
      Handler: 'index.handler',
      Runtime: 'nodejs12.x',
      MemorySize: 256,
      ReservedConcurrentExecutions: 20,
      Timeout: 20,
    })
  )
})

test('events have been added', () => {
  const app = new cdk.App()
  const stack = new S3ThumbnailStack(app, stackName)

  expectCDK(stack).to(haveResourceLike('AWS::Lambda::EventSourceMapping'))

  expectCDK(stack).to(
    haveResourceLike('Custom::S3BucketNotifications', {
      BucketName: {},
      NotificationConfiguration: {
        QueueConfigurations: [
          {
            Events: ['s3:ObjectCreated:*'],
            Filter: {
              Key: {
                FilterRules: [
                  {
                    Name: 'prefix',
                    Value: 'photos/',
                  },
                ],
              },
            },
          },
        ],
      },
    })
  )
})
