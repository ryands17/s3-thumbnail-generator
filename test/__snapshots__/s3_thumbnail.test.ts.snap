// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot works correctly 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584bArtifactHash6D2CA43C": Object {
      "Description": "Artifact hash for asset \\"20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584b\\"",
      "Type": "String",
    },
    "AssetParameters20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584bS3BucketD5770D8A": Object {
      "Description": "S3 bucket for asset \\"20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584b\\"",
      "Type": "String",
    },
    "AssetParameters20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584bS3VersionKey7334E3C3": Object {
      "Description": "S3 key for asset version \\"20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584b\\"",
      "Type": "String",
    },
    "AssetParametersc509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00ArtifactHash1887B8B1": Object {
      "Description": "Artifact hash for asset \\"c509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00\\"",
      "Type": "String",
    },
    "AssetParametersc509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00S3Bucket734CD5E5": Object {
      "Description": "S3 bucket for asset \\"c509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00\\"",
      "Type": "String",
    },
    "AssetParametersc509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00S3VersionKey4B8EB21C": Object {
      "Description": "S3 key for asset version \\"c509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": Object {
      "DependsOn": Array [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = (event, context) => {
    // eslint-disable-next-line @typescript-eslint/no-require-imports, import/no-extraneous-dependencies
    const s3 = new (require('aws-sdk').S3)();
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const https = require('https');
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const url = require('url');
    log(JSON.stringify(event, undefined, 2));
    const props = event.ResourceProperties;
    if (event.RequestType === 'Delete') {
        props.NotificationConfiguration = {}; // this is how you clean out notifications
    }
    const req = {
        Bucket: props.BucketName,
        NotificationConfiguration: props.NotificationConfiguration,
    };
    return s3.putBucketNotificationConfiguration(req, (err, data) => {
        log({ err, data });
        if (err) {
            return submitResponse('FAILED', err.message + \`\\\\nMore information in CloudWatch Log Stream: \${context.logStreamName}\`);
        }
        else {
            return submitResponse('SUCCESS');
        }
    });
    function log(obj) {
        console.error(event.RequestId, event.StackId, event.LogicalResourceId, obj);
    }
    // eslint-disable-next-line max-len
    // adapted from https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html#cfn-lambda-function-code-cfnresponsemodule
    // to allow sending an error messge as a reason.
    function submitResponse(responseStatus, reason) {
        const responseBody = JSON.stringify({
            Status: responseStatus,
            Reason: reason || 'See the details in CloudWatch Log Stream: ' + context.logStreamName,
            PhysicalResourceId: event.PhysicalResourceId || event.LogicalResourceId,
            StackId: event.StackId,
            RequestId: event.RequestId,
            LogicalResourceId: event.LogicalResourceId,
            NoEcho: false,
        });
        log({ responseBody });
        const parsedUrl = url.parse(event.ResponseURL);
        const options = {
            hostname: parsedUrl.hostname,
            port: 443,
            path: parsedUrl.path,
            method: 'PUT',
            headers: {
                'content-type': '',
                'content-length': responseBody.length,
            },
        };
        const request = https.request(options, (r) => {
            log({ statusCode: r.statusCode, statusMessage: r.statusMessage });
            context.done();
        });
        request.on('error', (error) => {
            log({ sendError: error });
            context.done();
        });
        request.write(responseBody);
        request.end();
    }
};",
        },
        "Description": "AWS CloudFormation handler for \\"Custom::S3BucketNotifications\\" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": Array [
          Object {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584bS3BucketD5770D8A",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584bS3VersionKey7334E3C3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters20e06317b2e4aa8c9a1052d2e0f40d4c738bc91399a9ea24aa2b1e013bfe584bS3VersionKey7334E3C3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allImagesBucket8FED7BDD": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "all-images-bucket",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "allImagesBucketNotificationsB7557D02": Object {
      "DependsOn": Array [
        "thumbnailQueuePolicy96FDE79E",
        "thumbnailQueue67869867",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "allImagesBucket8FED7BDD",
        },
        "NotificationConfiguration": Object {
          "QueueConfigurations": Array [
            Object {
              "Events": Array [
                "s3:ObjectCreated:*",
              ],
              "Filter": Object {
                "Key": Object {
                  "FilterRules": Array [
                    Object {
                      "Name": "prefix",
                      "Value": "photos/",
                    },
                  ],
                },
              },
              "QueueArn": Object {
                "Fn::GetAtt": Array [
                  "thumbnailQueue67869867",
                  "Arn",
                ],
              },
            },
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "resizeImageE8C4D857": Object {
      "DependsOn": Array [
        "resizeImageServiceRoleDefaultPolicy39E5149B",
        "resizeImageServiceRole7BA22CA0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00S3Bucket734CD5E5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00S3VersionKey4B8EB21C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc509aa98d9ff2fccd7449312e4b4de3f27ecd3c052473389af859a6770df8a00S3VersionKey4B8EB21C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "PREFIX": "photos/",
            "QUEUE_URL": Object {
              "Ref": "thumbnailQueue67869867",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "ReservedConcurrentExecutions": 20,
        "Role": Object {
          "Fn::GetAtt": Array [
            "resizeImageServiceRole7BA22CA0",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 20,
      },
      "Type": "AWS::Lambda::Function",
    },
    "resizeImageLogRetention12639EDD": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "resizeImageE8C4D857",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "resizeImageServiceRole7BA22CA0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "resizeImageServiceRoleDefaultPolicy39E5149B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:PutObject",
                "s3:GetObject",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "allImagesBucket8FED7BDD",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "thumbnailQueue67869867",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "resizeImageServiceRoleDefaultPolicy39E5149B",
        "Roles": Array [
          Object {
            "Ref": "resizeImageServiceRole7BA22CA0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "resizeImageSqsEventSourceS3ThumbnailStackthumbnailQueue5D2B250932806F5A": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "thumbnailQueue67869867",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "resizeImageE8C4D857",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "thumbnailQueue67869867": Object {
      "Properties": Object {
        "MessageRetentionPeriod": 172800,
        "QueueName": "thumbnailPayload",
        "ReceiveMessageWaitTimeSeconds": 20,
        "VisibilityTimeout": 60,
      },
      "Type": "AWS::SQS::Queue",
    },
    "thumbnailQueuePolicy96FDE79E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Condition": Object {
                "ArnLike": Object {
                  "aws:SourceArn": Object {
                    "Fn::GetAtt": Array [
                      "allImagesBucket8FED7BDD",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "s3.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "thumbnailQueue67869867",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "thumbnailQueue67869867",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
  },
}
`;
